pkg/docker/fast_push.go

```diff
@@ -148,7 +183,76 @@ func uploadFile(objectType string, digest string, path string, token string) err
 		return errors.New("Bad response: " + strconv.Itoa(resp.StatusCode))
 	}
 
+	// Decode the JSON payload
+	decoder := json.NewDecoder(resp.Body)
+	var data S3Config
+	err = decoder.Decode(&data)
+	if err != nil {
+		return err
+	}
+
+	// Open the file for uploading
+	file, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	// Upload the file using an S3 client
 	console.Debug("multi-part uploading file: " + path)
+	cfg := aws.NewConfig()
+	cfg.BaseEndpoint = &data.Endpoint
+	cfg.Credentials = credentials.StaticCredentialsProvider{
+		Value: aws.Credentials{
+			AccessKeyID:     data.AccessKeyId,
+			SecretAccessKey: data.SecretAccessKey,
+			SessionToken:    data.SecretAccessKey,
+			Expires:         time.Unix(data.Expires, 0),
+		},
+	}
+	s3Client := s3.NewFromConfig(*cfg)
+	uploadParams := &s3.PutObjectInput{
+		Bucket: aws.String(data.Bucket),
+		Key:    aws.String(data.Key),
+		Body:   file,
+	}
+	_, err = s3Client.PutObject(context.Background(), uploadParams)
+	if err != nil {
+		return err
+	}
```

Comment:
This will need to be changed to use the [managed uploader](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/feature/s3/manager#NewUploader) since:
- files will often be over 5GB
- the uploader takes care of retry / is much more resilient

Reply:
Resolved in [8809dea](https://github.com/replicate/cog/pull/2128/commits/8809deac27ece2069bad51328b1f4fefab75ff6c)